### PR TITLE
Geometric VOF using piecewise linear interface calculation (PLIC) and a split Lagrangian (onto-cell) advection scheme 

### PR DESCRIPTION
**Pull request by gdeskos**

### CODE REVIEW

File: amr-wind/equation_systems/vof/vof_advection.H [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                mfi_info.EnableTiling(amrex::IntVect(1024, 1024, 1024))
+                    .SetDynamic(true);
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+            for (amrex::MFIter mfi(dof_field(lev), mfi_info); mfi.isValid();
+                 ++mfi) {
+                const auto& bx = mfi.tilebox();
+                amrex::FArrayBox tmpfab(amrex::grow(bx, 1), 3 * VOF::ndim);
+                tmpfab.setVal(0.0);
```
Comment (michaeljbrazell) [inline amr-wind/equation_systems/vof/vof_advection.H:55]: maybe you need this: tmpfab.setval<amrex::RunOn::Device>(0.0); ?

Comment (gdeskos) [reply amr-wind/equation_systems/vof/vof_advection.H]: Done !

File: amr-wind/core/SimTime.cpp [hunk 2]
```diff
@@ -92,7 +92,8 @@ void SimTime::set_current_cfl(
     amrex::Real dt_new = 2.0 * m_max_cfl / cfl_unit_time;
 
     // Restrict timestep during initialization phase
-    if (m_is_init) dt_new *= m_init_shrink;
+    if (m_is_init)
+        dt_new = amrex::min(m_init_shrink * dt_new, std::abs(m_fixed_dt));
```
Comment (sayerhs) [inline amr-wind/core/SimTime.cpp:96]: Remove this as this was a hack to allow you to run the dam break problem. A better fix has been added to the main branch.

File: amr-wind/equation_systems/vof/SplitAdvection.H [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#ifndef SplitAdvection_H
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/SplitAdvection.H:1]: Seems like we can combine `lagrangian_advection.H` and `SplitAdvection.H` into a single header.

File: amr-wind/equation_systems/vof/SplitAdvection.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    Box const& domain = geom[lev].Domain();
+    const auto domlo = amrex::lbound(domain);
+    const auto domhi = amrex::ubound(domain);
+
+    Array4<Real> fluxL = makeArray4(p, bxg1, 1);
+    p += fluxL.size();
+    Array4<Real> fluxC = makeArray4(p, bxg1, 1);
+    p += fluxC.size();
+    Array4<Real> fluxR = makeArray4(p, bxg1, 1);
+    p += fluxR.size();
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/SplitAdvection.cpp:41]: `clang-tidy` warning here is relevant.

File: amr-wind/equation_systems/vof/SplitAdvection.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    if (isweep % 3 == 0) {
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/SplitAdvection.cpp:43]: This code duplication can be collapsed into a single function.

File: amr-wind/equation_systems/vof/vof_advection.H [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+#define VOF_ADVECTION_H
+
+#include "amr-wind/equation_systems/vof/vof.H"
+#include "amr-wind/equation_systems/vof/SplitAdvection.H"
+
+namespace amr_wind {
+namespace pde {
+
+/** Godunov scheme for VOF
+ *  \ingroup icns
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_advection.H:12]: Should not belong to `ICNS` documentation group.

File: amr-wind/equation_systems/vof/vof_advection.H [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        auto& dof_field = fields.field;
+        auto bcrec_device = dof_field.bcrec_device();
+        //
+        // Advect volume using the Explicit Lagrangian onto-cell Sweeping method
+        // with PLIC reconstruction
+        //
+
+        // Define the sweep time
+        isweep += 1;
+        if (isweep > 3) isweep = 1;
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_advection.H:41]: Why is this needed?

File: amr-wind/equation_systems/vof/vof_ops.H [hunk 8]
```diff
@@ -32,7 +32,7 @@ struct FieldRegOp<VOF, Scheme>
         auto& iomgr = sim.io_manager();
         iomgr.register_io_var(fields.field.name());
         iomgr.register_io_var(levelset.name());
-        iomgr.register_io_var(normal.name());
+        // iomgr.register_io_var(normal.name());
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_ops.H:35]: Remove

File: amr-wind/equation_systems/vof/vof_ops.H [hunk 9]
```diff
@@ -32,7 +32,7 @@ struct FieldRegOp<VOF, Scheme>
         auto& iomgr = sim.io_manager();
         iomgr.register_io_var(fields.field.name());
         iomgr.register_io_var(levelset.name());
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_ops.H:34]: Are these auxiliary variables required for restart? If not, I would recommend using `iomgr.register_output_var` for these aux variables.

File: amr-wind/equation_systems/vof/vof_ops.H [hunk 10]
```diff
@@ -48,71 +48,9 @@ struct ComputeRHSOp<VOF, Scheme>
...
-                        // phi_nph(i, j, k) =
-                        //     0.5 * (phi(i, j, k) + phi_o(i, j, k));
-                    });
-            }
-        }
-    }
+    void corrector_rhs(const DiffusionType, const amrex::Real) {}
 
     // data members
     PDEFields& fields;
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/vof_ops.H:56]: Unused variable, can be eliminated.

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 11]
```diff
@@ -1,34 +1,249 @@
...
+// the gradient is computed with a multiplicative factor of -32:
+//  mm = - 32 * grad (c)
+//----------------------------------------------------------------
+//
+// Known problems: the index (1,1,1), i.e. the central cell in the block, never
+occurs:
+// Therefore an isolated droplet will have a normal with all components to zero.
+// Translated into f90 by Stephane Z.
+*/
+
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:21]: Prefer to use doxygen style comment here. 

```suggestion
/** Young's finite-difference gradient scheme.
 *
 *  The gradient is computed with a multiplicative factor of -32:
 *  \f[ mm = - 32 * \grad (c) \f]
 *
 *  Known problems: the index (1,1,1), i.e. the central cell in the block, never
 *  occurs: Therefore an isolated droplet will have a normal with all components
 *  to zero.
 */
```

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 12]
```diff
@@ -1,34 +1,249 @@
...
+         volfrac(i - 1, j - 1, k) + volfrac(i - 1, j + 1, k) +
+         volfrac(i - 1, j, k);
+    m2 = volfrac(i + 1, j, k - 1) + volfrac(i + 1, j, k + 1) +
+         volfrac(i + 1, j - 1, k) + volfrac(i + 1, j + 1, k) +
+         volfrac(i + 1, j, k);
+    if (m1 > m2) {
+        m[0][0] = 1.0;
+    } else {
+        m[0][0] = -1.0;
+    }
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:93]:
```suggestion
    m[0][0] = (m1 > m2) ? 1.0 : -1.0;
```

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 13]
```diff
@@ -1,34 +1,249 @@
...
+         volfrac(i, j - 1, k);
+    m2 = volfrac(i, j + 1, k - 1) + volfrac(i, j + 1, k + 1) +
+         volfrac(i + 1, j + 1, k) + volfrac(i - 1, j + 1, k) +
+         volfrac(i, j + 1, k);
+
+    if (m1 > m2) {
+        m[1][1] = 1.;
+    } else {
+        m[1][1] = -1.;
+    }
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:127]:
```suggestion
    m[1][1] = (m1 > m2) ? 1.0 : -1.0;
```

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 14]
```diff
@@ -1,34 +1,249 @@
...
+         volfrac(i, j, k - 1);
+    m2 = volfrac(i - 1, j, k + 1) + volfrac(i + 1, j, k + 1) +
+         volfrac(i, j - 1, k + 1) + volfrac(i, j + 1, k + 1) +
+         volfrac(i, j, k + 1);
+
+    if (m1 > m2) {
+        m[2][2] = 1.;
+    } else {
+        m[2][2] = -1.;
+    }
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:161]:
```suggestion
    m[2][2] = (m1 > m2) ? 1.0 : -1.0;
```

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 15]
```diff
@@ -1,34 +1,249 @@
...
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void mixed_youngs_central_normal(
+    int i,
+    int j,
+    int k,
+    amrex::Array4<amrex::Real const> const& volfrac,
+    amrex::Real& mx,
+    amrex::Real& my,
+    amrex::Real& mz) noexcept
+{
+    amrex::Real m[AMREX_SPACEDIM + 1][AMREX_SPACEDIM];
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:79]: Is this supposed to be a vector? If so 
```suggestion
   vs::Vector mvec[AMREX_SPACEDIM+1];
```

File: amr-wind/equation_systems/vof/volume_fractions.H [hunk 16]
```diff
@@ -1,34 +1,249 @@
...
+    m[1][0] = m[1][0] / t0;
+    m[1][1] = m[1][1] / t0;
+    m[1][2] = m[1][2] / t0;
+
+    t0 = amrex::Math::abs(m[2][0]) + amrex::Math::abs(m[2][1]) +
+         amrex::Math::abs(m[2][2]);
+    m[2][0] = m[2][0] / t0;
+    m[2][1] = m[2][1] / t0;
+    m[2][2] = m[2][2] / t0;
+
```
Comment (sayerhs) [inline amr-wind/equation_systems/vof/volume_fractions.H:183]: Replace with `vs::Vector` and use 

```C++
for (int i=0; i < AMREX_SPACEDIM; ++i) m[i].normalize();
```

File: amr-wind/physics/multiphase/MultiPhase.H [hunk 17]
```diff
@@ -14,7 +14,7 @@
  */
 namespace amr_wind {
 
-enum class InterfaceTrackingMethod { VOF, LS };
+enum class InterfaceCapturingMethod { VOF, LS };
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.H:17]:
```suggestion
enum class InterfaceCapturingMethod 
{
    VOF,  ///< Volume of fluid 
    LS.   ///< Levelset
};
```

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 18]
```diff
@@ -13,59 +15,127 @@ MultiPhase::MultiPhase(CFDSim& sim)
...
 {
-    return m_interface_tracking_method;
+    return m_interface_capturing_method;
 }
 
 void MultiPhase::post_init_actions()
 {
 
-    if (m_interface_tracking_method == InterfaceTrackingMethod::VOF) {
+    if (m_interface_capturing_method == InterfaceCapturingMethod::VOF) {
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:60]: `switch` instead?

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 19]
```diff
@@ -13,59 +15,127 @@ MultiPhase::MultiPhase(CFDSim& sim)
...
+        m_total_volfrac = volume_fraction_sum();
+        const auto& geom = m_sim.mesh().Geom();
+        const amrex::Real total_vol = geom[0].ProbDomain().volume();
+        amrex::Print() << "Volume of Fluid diagnostics:" << std::endl;
+        amrex::Print() << "   Water Volume Fractions Sum : " << m_total_volfrac
+                       << std::endl;
+        amrex::Print() << "   Air Volume Fractions Sum : "
+                       << total_vol - m_total_volfrac << std::endl;
+        amrex::Print() << " " << std::endl;
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:88]: Protect print stuff with `verbose` or remove it.

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 20]
```diff
@@ -13,59 +15,127 @@ MultiPhase::MultiPhase(CFDSim& sim)
...
+{
+    using namespace amrex;
+    BL_PROFILE("amr-wind::multiphase::ComputeVolumeFractionSum");
+    const int nlevels = m_sim.repo().num_active_levels();
+    const auto& geom = m_sim.mesh().Geom();
+    auto& mesh = m_sim.mesh();
+    const auto grids = mesh.boxArray();
+    const auto dmap = mesh.DistributionMap();
+
+    amrex::Real TotalVolumeFrac = 0.0;
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:102]:
```suggestion
    amrex::Real total_volume_frac = 0.0;
```

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 21]
```diff
@@ -13,59 +15,127 @@ MultiPhase::MultiPhase(CFDSim& sim)
...
+                                     geom[lev].CellSize()[1] *
+                                     geom[lev].CellSize()[2];
+
+        TotalVolumeFrac += amrex::ReduceSum(
+            vof, level_mask, 0,
+            [=] AMREX_GPU_HOST_DEVICE(
+                amrex::Box const& bx,
+                amrex::Array4<amrex::Real const> const& volfrac,
+                amrex::Array4<int const> const& mask_arr) -> amrex::Real {
+                amrex::Real Vol_Fab = 0.0;
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:129]:
```suggestion
                amrex::Real vol_fab = 0.0;
```

File: amr-wind/physics/multiphase/MultiPhase.cpp [hunk 22]
```diff
@@ -135,7 +205,10 @@ void MultiPhase::levelset2vof()
     const int nlevels = m_sim.repo().num_active_levels();
     const auto& geom = m_sim.mesh().Geom();
 
+    const amrex::Real normal_default = 0.0;
     auto& normal = m_sim.repo().get_field("interface_normal");
+    BCScalar bc_normal(normal);
+    bc_normal(normal_default);
```
Comment (sayerhs) [inline amr-wind/physics/multiphase/MultiPhase.cpp:211]:
```suggestion
    normal.set_default_fillpatch_bc(m_sim.time());
```

This should be done once when you declare field, not every time this method is called.

File: amr-wind/equation_systems/vof/SplitAdvection.H (hunk 3 above)
Comment (gdeskos) [reply amr-wind/equation_systems/vof/SplitAdvection.H:1]: Sure, but I will also add a eulerian_advection file so I prefer to keep them separate

File: amr-wind/equation_systems/vof/vof_advection.H (hunk 7 above)
Comment (gdeskos) [reply amr-wind/equation_systems/vof/vof_advection.H:41]: Well it is not needed it can simply be isweep=isweep%3

File: amr-wind/equation_systems/vof/vof_ops.H (hunk 9 above)
Comment (gdeskos) [reply amr-wind/equation_systems/vof/vof_ops.H:34]: Done

File: amr-wind/equation_systems/vof/volume_fractions.H (hunk 16 above)
Comment (gdeskos) [reply amr-wind/equation_systems/vof/volume_fractions.H:183]: Not sure what vs::Vector is it different to the std::vector ?

File: amr-wind/physics/multiphase/MultiPhase.cpp (hunk 19 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/MultiPhase.cpp:88]:
```suggestion
    // Compute the print the total volume fraction
    if (m_interface_capturing_method == InterfaceCapturingMethod::VOF &&
        m_verbose > 0) {
        m_total_volfrac = volume_fraction_sum();
        const auto& geom = m_sim.mesh().Geom();
        const amrex::Real total_vol = geom[0].ProbDomain().volume();
        amrex::Print() << "Volume of Fluid diagnostics:" << std::endl;
        amrex::Print() << "   Water Volume Fractions Sum : " << m_total_volfrac
                       << std::endl;
        amrex::Print() << "   Air Volume Fractions Sum : "
                       << total_vol - m_total_volfrac << std::endl;
        amrex::Print() << " " << std::endl;
    }
```

File: amr-wind/physics/multiphase/MultiPhase.cpp (hunk 18 above)
Comment (gdeskos) [reply amr-wind/physics/multiphase/MultiPhase.cpp:60]: does it make any difference ?

File: amr-wind/equation_systems/vof/volume_fractions.H (hunk 15 above)
Comment (gdeskos) [reply amr-wind/equation_systems/vof/volume_fractions.H:79]: it should be a 4X3 array

**Comment by sayerhs** [comment]
Closing in favor of #308
